From 3c17b220d9ababfcf874f1c253894272d4b120e1 Mon Sep 17 00:00:00 2001
From: Egor Kutovoi <e.kutovoi@partner.samsung.com>
Date: Mon, 20 Mar 2023 12:43:32 +0300
Subject: [PATCH 110/135] [cmake] Don't specify axdimm includes, use build dir

After adding headers to public dependencies in cmake we can now omit all
`target_include_directories` statements and only link with the required
libraries.

Considering this, we can also don't pass the full cmake root directory
enviroment variable and pass only a path to the build directory. This
way we remove unnecessary dependencies and enable further improvements.

Related to: AXDIMM-638
---
 CMakeLists.txt           |  2 +-
 caffe2/CMakeLists.txt    | 10 +---------
 cmake/Dependencies.cmake |  5 ++---
 setup.py                 | 10 +++++-----
 4 files changed, 9 insertions(+), 18 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 36a3c36a..7c73d2b9 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -192,7 +192,7 @@ option(USE_COLORIZE_OUTPUT "Colorize output during compilation" ON)
 option(USE_ASAN "Use Address+Undefined Sanitizers" OFF)
 option(USE_TSAN "Use Thread Sanitizer" OFF)
 option(USE_AXDIMM "Use AXDIMM" OFF)
-option(CMAKE_AXDIMM_ROOT_PATH "A path to AXDIMM library root directory" "")
+option(CMAKE_AXDIMM_BUILD_DIR "A path to AXDIMM library build output directory" "")
 option(USE_CUDA "Use CUDA" ON)
 cmake_dependent_option(
      BUILD_LAZY_CUDA_LINALG "Build cuda linalg ops as separate library" ON "USE_CUDA AND LINUX AND BUILD_PYTHON" OFF)
diff --git a/caffe2/CMakeLists.txt b/caffe2/CMakeLists.txt
index 956db7d6..f42e0cb6 100644
--- a/caffe2/CMakeLists.txt
+++ b/caffe2/CMakeLists.txt
@@ -105,13 +105,6 @@ if(INTERN_BUILD_ATEN_OPS)
   list(APPEND Caffe2_DEPENDENCY_INCLUDE ${ATen_THIRD_PARTY_INCLUDE})
 endif()
 
-if(USE_AXDIMM)
-  if(NOT CMAKE_AXDIMM_ROOT_PATH)
-    message(FATAL_ERROR "CMAKE_AXDIMM_ROOT_PATH is not specified, can't find AXDIMM artifacts!")
-  endif()
-  list(APPEND Caffe2_AXDIMM_INCLUDE "${CMAKE_AXDIMM_ROOT_PATH}")
-endif()
-
 # ---[ Caffe2 build
 # Note: the folders that are being commented out have not been properly
 # addressed yet.
@@ -1374,8 +1367,6 @@ target_include_directories(torch_cpu SYSTEM PRIVATE "${Caffe2_DEPENDENCY_INCLUDE
 
 if (USE_AXDIMM)
   target_link_libraries(torch_axdimm PRIVATE ${Caffe2_AXDIMM_DEPENDENCY_LIBS})
-
-  target_include_directories(torch_axdimm PRIVATE ${Caffe2_AXDIMM_INCLUDE})
 endif ()
 
 target_compile_options(torch_cpu PRIVATE "-DCAFFE2_BUILD_MAIN_LIB")
@@ -2008,6 +1999,7 @@ if(BUILD_PYTHON)
     target_include_directories(caffe2_pybind11_state_axdimm PRIVATE $<INSTALL_INTERFACE:include>)
     target_include_directories(caffe2_pybind11_state_axdimm PRIVATE ${Caffe2_CPU_INCLUDE} ${Caffe2_AXDIMM_INCLUDE})
     target_link_libraries(caffe2_pybind11_state_axdimm PRIVATE torch_library python::python pybind::pybind11)
+    target_link_libraries(caffe2_pybind11_state_axdimm PRIVATE Axdimm::AxdimmAI)
 
     if(USE_MKLDNN)
         target_link_libraries(caffe2_pybind11_state_axdimm PRIVATE caffe2::mkldnn)
diff --git a/cmake/Dependencies.cmake b/cmake/Dependencies.cmake
index 0409d468..99b1023e 100644
--- a/cmake/Dependencies.cmake
+++ b/cmake/Dependencies.cmake
@@ -1828,9 +1828,8 @@ add_subdirectory(${PROJECT_SOURCE_DIR}/third_party/fmt)
 # to define one, to stop cmake complaining.
 # Axdimm cmake files are required only to export axdimm library packages
 if(USE_AXDIMM)
-  include_directories(SYSTEM ${CMAKE_AXDIMM_ROOT_PATH})
-  list(APPEND CMAKE_PREFIX_PATH "${CMAKE_AXDIMM_ROOT_PATH}/build/ai/cmake")
-  list(APPEND CMAKE_PREFIX_PATH "${CMAKE_AXDIMM_ROOT_PATH}/build/secure/cmake")
+  list(APPEND CMAKE_PREFIX_PATH "${CMAKE_AXDIMM_BUILD_DIR}/ai/cmake")
+  list(APPEND CMAKE_PREFIX_PATH "${CMAKE_AXDIMM_BUILD_DIR}/secure/cmake")
   find_package(AxdimmAI REQUIRED)
   find_package(AxdimmSecure REQUIRED)
   set(AXDIMM_LIBS Axdimm::AxdimmSecure Axdimm::AxdimmAI)
diff --git a/setup.py b/setup.py
index a3de2171..7e74c542 100644
--- a/setup.py
+++ b/setup.py
@@ -17,8 +17,8 @@
 #   USE_AXDIMM=1
 #     enable AXDIMM build
 #
-#   CMAKE_AXDIMM_ROOT_PATH=<path>
-#     path to AXDIMM library root directectory
+#   CMAKE_AXDIMM_BUILD_DIR=<path>
+#     path to AXDIMM library build output directectory
 #
 #   CFLAGS
 #     flags to apply to both C and C++ files to be compiled (a quirk of setup.py
@@ -516,10 +516,10 @@ class build_ext(setuptools.command.build_ext.build_ext):
             report('-- Using AXDIMM build')
         else:
             report('-- Not using AXDIMM')
-        if cmake_cache_vars['CMAKE_AXDIMM_ROOT_PATH']:
-            report('-- Detected AXDIMM root dir at {}'.format(cmake_cache_vars['CMAKE_AXDIMM_ROOT_PATH']))
+        if cmake_cache_vars['CMAKE_AXDIMM_BUILD_DIR']:
+            report('-- Detected AXDIMM build dir at {}'.format(cmake_cache_vars['CMAKE_AXDIMM_BUILD_DIR']))
         else:
-            report('-- Not using CMAKE_AXDIMM_ROOT_PATH')
+            report('-- Not using CMAKE_AXDIMM_BUILD_DIR')
         if cmake_cache_vars['USE_MKLDNN']:
             report('-- Using MKLDNN')
             if cmake_cache_vars['USE_MKLDNN_ACL']:
-- 
2.34.1

